src/compiler/checker.ts

```diff
@@ -37389,6 +37399,79 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                if (singleReturn || !returnStatement.expression) return true;
+                singleReturn = returnStatement.expression;
+            });
+            if (bailedEarly || !singleReturn) return undefined;
+        }
+
+        return checkIfExpressionRefinesAnyParameter(singleReturn);
+
+        function checkIfExpressionRefinesAnyParameter(expr: Expression): TypePredicate | undefined {
+            expr = skipParentheses(expr, /*excludeJSDocTypeAssertions*/ true);
+            const returnType = checkExpressionCached(expr);
+            if (!(returnType.flags & TypeFlags.Boolean)) return undefined;
+
+            return forEach(func.parameters, (param, i) => {
+                const initType = getTypeOfSymbol(param.symbol);
+                if (!initType || initType.flags & TypeFlags.Boolean || !isIdentifier(param.name) || isSymbolAssigned(param.symbol)) {
+                    // Refining "x: boolean" to "x is true" or "x is false" isn't useful.
+                    return;
+                }
+                const trueType = checkIfExpressionRefinesParameter(expr, param, initType);
+                if (trueType) {
+                    return createTypePredicate(TypePredicateKind.Identifier, unescapeLeadingUnderscores(param.name.escapedText), i, trueType);
+                }
+            });
+        }
+
+        function checkIfExpressionRefinesParameter(expr: Expression, param: ParameterDeclaration, initType: Type): Type | undefined {
+            const baseAntecedent = (expr as Expression & { flowNode?: FlowNode; }).flowNode ||
+                expr.parent.kind === SyntaxKind.ReturnStatement && (expr.parent as ReturnStatement).flowNode ||
+                { flags: FlowFlags.Start };
+            const sharedAntecedent = { ...baseAntecedent, flags: baseAntecedent.flags & FlowFlags.Shared };
+            const trueCondition: FlowCondition = {
+                flags: FlowFlags.TrueCondition,
+                node: expr,
+                antecedent: sharedAntecedent,
+            };
+
+            const trueType = getFlowTypeOfReference(param.name, initType, initType, func, trueCondition);
+            // Since trueType is a subtype of initType, isTypeAssignableTo(initType, trueType) is a test for equivalence (modulo any types).
+            if (trueType === initType || ((initType.flags & TypeFlags.Any) ? (trueType.flags & TypeFlags.Any) : isTypeAssignableTo(initType, trueType))) {
+                return undefined;
+            }
+
+            // "x is T" means that x is T if and only if it returns true. If it returns false then x is not T.
+            // This means that if the function is called with an argument of type trueType, there can't be anything left in the `else` branch. It must reduce to `never`.
+            const falseCondition: FlowCondition = {
+                ...trueCondition,
+                flags: FlowFlags.FalseCondition,
+            };

```

Comment:
I don't know how engines optimize this today, but a couple years ago the spread operator was significantly slower than specifying all properties manually. Perhaps that would help here as well

```suggestion
            const falseCondition: FlowCondition = {
                flags: FlowFlags.FalseCondition,
                node: expr,
                antecedent: sharedAntecedent,
            };
```

Reply:
Additionally, `createFlowCondition` creates the object literal as follows:

```ts
return initFlowNode({ flags, antecedent, node: expression });
```

importantly, `antecedent` precedes `node`. By deviating from this ordering here a second hidden class is introduced for condition flow nodes, which may cause monomorphic accesses to turn polymorphic.

Reply:
The flow nodes are one of the places we haven't optimized yet; there's loads of out of order props that would benefit from consistent factories (it's just not that way yet)

Reply:
Ah, then it may not matter right now.